asset-registry/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+		#[pallet::weight(T::WeightInfo::register_asset())]
+		#[transactional]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: Option<T::AssetId>,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			let (asset_id, metadata) = T::OnRegisterAsset::process_asset(asset_id, metadata)?;
+
+			Self::insert_metadata(&asset_id, &metadata)?;
+
+			Self::deposit_event(Event::<T>::RegisteredAsset { asset_id, metadata });
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::update_asset())]
+		#[transactional]
+		pub fn update_asset(
+			origin: OriginFor<T>,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: T::AssetId,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			Self::update_metadata(&asset_id, &metadata)?;
+
+			Self::deposit_event(Event::<T>::UpdatedAsset { asset_id, metadata });
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::set_asset_location())]
+		#[transactional]
+		pub fn set_asset_location(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			location: Box<VersionedMultiLocation>,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			let old_metadata = Metadata::<T>::get(&asset_id).ok_or(Error::<T>::AssetNotFound)?;
+			let new_metadata = AssetMetadata {
+				location: Some((*location).clone()),
+				..old_metadata
+			};
+			Self::update_metadata(&asset_id, &new_metadata)?;
+
+			Self::deposit_event(Event::<T>::SetLocation { asset_id, location });
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn update_metadata(
+		asset_id: &T::AssetId,
+		metadata: &AssetMetadata<T::Balance, T::CustomMetadata>,
+	) -> DispatchResult {
+		let old_metadata = Metadata::<T>::get(&asset_id).ok_or(Error::<T>::AssetNotFound)?;
+
+		if let Some(location) = old_metadata.location {
+			// remove the old location lookup
+			let location: MultiLocation = location.try_into().map_err(|()| Error::<T>::BadVersion)?;
+			MultiLocationLookup::<T>::remove(location);
+		}
+
+		Self::insert_metadata(asset_id, metadata)?;
+
+		Ok(())
+	}
+
+	pub fn fetch_metadata_by_location(
+		location: &MultiLocation,
+	) -> Option<AssetMetadata<T::Balance, T::CustomMetadata>> {
+		let asset_id = MultiLocationLookup::<T>::get(location)?;
+		Metadata::<T>::get(asset_id)
+	}
+
+	fn insert_metadata(
+		asset_id: &T::AssetId,
+		metadata: &AssetMetadata<T::Balance, T::CustomMetadata>,
+	) -> DispatchResult {
+		// if the metadata contains a location, set the MultiLocationLookup
+		if let Some(location) = metadata.location.clone() {
+			let location: MultiLocation = location.try_into().map_err(|()| Error::<T>::BadVersion)?;
+			ensure!(
+				!MultiLocationLookup::<T>::contains_key(&location),
+				Error::<T>::ConflictingLocation
+			);
+			MultiLocationLookup::<T>::insert(location, asset_id);
+		}
+
+		Metadata::<T>::insert(&asset_id, &metadata);
```

Comment:
We could add a judegment before insert operation, if asset_id exist, throw AssetIdExist Error.

Reply:
I'm not sure `try_mutate` or `insert` which one is best:
```
Metadata::<T>::try_mutate(
	&asset_id,
	|maybe_asset_metadatas| -> DispatchResult {
		ensure!(maybe_asset_metadatas.is_none(), Error::<T>::AssetIdExisted);

		*maybe_asset_metadatas = Some(metadata.clone());
		Ok(())
	},
)?;

Ok(())
```

Reply:
I added the check in `register_asset` - I named the error `ConflictingAssetId`, to be consistent with the existing `ConflictingLocation` error.

I also updated the code to use try_mutate where possible. Instead of `AssetIdNotExisted` I'm reusing the existing `AssetNotFound`